presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java

```diff
@@ -764,34 +780,100 @@ public Page getNextPage()
                 if (currentPosition == totalRowCount) {
                     return null;
                 }
                 if (!advanceToNextRowGroup()) {
+                    if (reorderFilters && currentRowGroup < ROW_GROUP_REVIEW_INTERVAL * 2) {
+                        // If file ends before 2 reorder checks, do
+                        // one more. Subsequent splits may start with
+                        // the adaptation from the previous one.
+                        reader.maybeReorderFilters();
+                    }
                     filePosition = fileRowCount;
                     currentPosition = totalRowCount;
                     return resultPage();
                 }
                 if (qualifyingSet == null) {
                     qualifyingSet = new QualifyingSet();
                 }
-                qualifyingSet.setRange((int) currentGroupRowCount);
+                qualifyingSet.setRange(0, Math.min(ariaBatchRows, currentGroupRowCount));
                 reader.setQualifyingSets(qualifyingSet, null);
-                if (reorderFilters && (currentRowGroup & 0x3) != 0 && currentRowGroup != 0) {
-                    // Reconsider filter order every 4 row groups.
-                    reader.maybeReorderFilters();
+                if (currentRowGroup != 0 && (currentRowGroup % ROW_GROUP_REVIEW_INTERVAL == 0)) {
+                    // Decay row size stats and reconsider filter
+                    // order every ROW_GROUP_REVIEW_INTERVAL row
+                    // groups.
+                    numAriaRows /= 2;
+                    numAriaBytes /= 2;
+                    if (reorderFilters) {
+                        reader.maybeReorderFilters();
+                    }
+                }
                 }
+            if (qualifyingSet.isEmpty()) {
+                qualifyingSet.setRange(qualifyingSet.getEnd(), Math.min(qualifyingSet.getEnd() + ariaBatchRows, currentGroupRowCount));
             }
-            qualifyingSet.clearTruncationPosition();
-            reader.advance();
+            int bytesBeforeAdvance = reader.getResultSizeInBytes();
+            int numResultsBeforeAdvance = reader.getNumResults();
+            try {
+                reader.advance();
+            }
+            catch (BatchTooLargeException e) {
+                // The reader ran out of budget. Retry with a smaller
+                // input qualifying set. First remove any values added
+                // by the last advance.  The input qualifying set is
+                // unaltered when the reader throws this exception.
+                // If we are at minimum batch size we must also have unlimited budget. Check that this path is not taken with minimum batch size.
+                verify(qualifyingSet.getEnd() - qualifyingSet.getPositions()[0] > MIN_BATCH_ROWS, "Running out of budget with minimum batch size");
```

Comment:
I find this message confusing. To me it says that minimum batch size is still too large, however, it really means that we ran out of *unlimited* budget. How about "Ran out of unlimited budget"?